### ISSUE TITLE
bug with filtering

### ISSUE THREAD

**Issue by chovy** [aurelia/framework#123]
I've managed to reproduce this infinite DOM rendering bug in a branch on the skeleton app when using filtering:

Here is a working example with the bug
https://github.com/chovy/skeleton-navigation/tree/bug/bad-filtering

Clone and switch to that branch, then run `gulp watch`

1. Goto flickr page
2. type 'view' or any title string to match on
3. inspect DOM in chrome
4. Notice it is constantly being updated
5. also see console log

You will see the DOM is infinitely re-rendering.

I copied this example from the documentation here: http://aurelia.io/docs.html#html-extensions

Here is a video demonstrating the issue:
https://www.dropbox.com/s/bnmteb1u2nwxuw9/skeleton-infinite-filter.mov?dl=0

@EisenbergEffect Hey Rob, you asked if I could re-create this in skeleton and was able to (see my branch above)

**Comment by Mordred** [comment]
Hi, this is happening because you are always returning new array instance. Basically `vm.filteredImages !== vm.filteredImages`. Try something like this:

```js
@computedFrom('images', 'searchText')
get filteredImages() { ... }
```

or with more code without `computedFrom`:

```js
  set searchText(value) {
    if (value !== this._searchText) {
      this._filteredImages = null;
      this._searchText = value;
    }
  }

  get searchText() {
    return this._searchText;
  }

  get filteredImages(){
    if ( this.searchText === '' ) return this.images;

    if (this._filteredImages) return this._filteredImages;

    return this._filteredImages = this.images.filter(item => {
        return item.title.toLowerCase().indexOf(this.searchText.toLowerCase()) !== -1;
    });
  }
```

Unfortunately there bug with `repeat`. When you type too fast, some elements will stay on the page, because `repeat` will loose reference to them.

**Comment by chovy** [comment]
This seems a bit unintuitive. But I guess it will have to do for now.

**Comment by jdanyow** [comment]
@chovy - @Mordred is correct.  The best-practices around computed array props are also discussed here:  http://aurelia.io/docs.html#adaptive-binding

Here's another approach you might find more intuitive:

Add this converter to the standard  [`flicker.js`](https://github.com/aurelia/skeleton-navigation/blob/master/src/flickr.js) included in skeleton-navigation:

```javascript
export class TitleFilterValueConverter {
  toView(images, searchText) {
    searchText = searchText.toLowerCase();
    return images.filter(item => item.title.toLowerCase().indexOf(searchText) !== -1);
 }
}
```

`ref` your search `<input>` which will enable you to pass it's value to the converter used in the images `repeat.for` binding:

```html
<template>
  <section class="au-animate">
    <h2>${heading}</h2>
    <div class="form-group">
      <label class="control-label" for="search">Search</label>
      <input class="form-control input-sm" type="text" id="search" value.bind="searchText" ref="searchInput">
    </div>
    <div class="row au-stagger">
        <div repeat.for="image of images | titleFilter:searchInput.value" class="col-sm-6 col-md-3 flickr-img au-animate">
          <a class="thumbnail">
            <img src.bind="image.media.m"/><br>
            <small>${image.title}</small>
          </a>
        </div>
      </div>
  </section>
</template>
```

**Comment by chovy** [comment]
thanks guys. both these solutions work, except in my real example I have to instantiate a video player.

I do that by calling: `new Clappr.Player()` for each video on the page. However as soon as the filter is run, these video players are lost.

I'm currently doing it in `attached()` which doesn't seem to get called for filters.

I tried this, but it doesn't work either:

```
export class TitleFilterValueConverter {
  toView(channels, searchText) {
    searchText = searchText.toLowerCase();
    let _filteredChannels = channels.filter(item => item.name.toLowerCase().indexOf(searchText) !== -1);

    ChannelList.getPlayers();
    return _filteredChannels;
  }
}
```

But this seems to be too soon. I need to create the video player only *after* the `<video>` tag has been added to the dom.
